cpp/include/raft/neighbors/detail/ivf_flat_search-inl.cuh

```diff
@@ -213,22 +213,33 @@ inline void search(raft::device_resources const& handle,
                "n_probes (number of clusters to probe in the search) must be positive.");
   auto n_probes = std::min<uint32_t>(params.n_probes, index.n_lists());
 
-  auto pool_guard = raft::get_pool_memory_resource(mr, n_queries * n_probes * k * 16);
+  // a batch size heuristic: try to keep the workspace within the specified size
+  constexpr uint32_t kExpectedWsSize = 1024 * 1024 * 1024;
+  const uint32_t max_queries =
+    std::min<uint32_t>(n_queries,
+                       raft::div_rounding_up_safe<uint64_t>(
+                         kExpectedWsSize, 16ull * uint64_t{n_probes} * k + 4ull * index.dim()));
```

Comment:
Why 16 and 4? Are these related to the date / index types?

Reply:
Yes, that's very approximate heuristic for the allocated buffers inside search_impl<>. I figured there's no point in calculating these precise because (1) batching it does not seem to affect performance much (2) we plan to work on our workspace size management soon anyway.

Reply:
Sounds good, thanks!